### PR TITLE
[MI-2639]: Fixed some issues and enhancement.

### PR DESCRIPTION
**Pull request by Nityanand13**
1. Fixed issue: [#116](https://github.com/mattermost/mattermost-plugin-servicenow/issues/116).
2. Fixed issue: [#114](https://github.com/Brightscout/mattermost-plugin-servicenow/issues/114).
3. Modified an error when a user deletes a deleted subscription.

Screenshot from 2023-02-01 13-12-52

### CODE REVIEW

File: webapp/src/containers/addOrEditSubscriptions/subComponents/index.tsx [hunk 1]
```diff
@@ -285,6 +285,13 @@ const AddOrEditSubscription = ({open, close, subscriptionData}: AddOrEditSubscri
         return Constants.SubscriptionAddedMsg;
     }, [apiError, apiResponseValid, subscriptionData]);
 
+    const [subscriptionNumber, setSubscriptionNumber] = useState('');
+    const getSubscriptionNumber = (number: string) => {
+        if (number) {
+            setSubscriptionNumber(number);
+        }
+    };
```
Comment (ayusht2810) [inline webapp/src/containers/addOrEditSubscriptions/subComponents/index.tsx:293]: can't we simply send `setSubscriptionNumber` as a prop?

File: webapp/src/containers/addOrEditSubscriptions/subComponents/index.tsx [hunk 2]
```diff
@@ -323,6 +331,7 @@ const AddOrEditSubscription = ({open, close, subscriptionData}: AddOrEditSubscri
             subscription_events: subscriptionEvents.join(','),
             channel_id: channel as string,
             sys_id: subscriptionData?.id as string,
+            subscription_number: subscriptionNumber,
```
[the remark this answers is not in the document]
Comment (Nityanand13) [reply webapp/src/containers/addOrEditSubscriptions/subComponents/index.tsx:334]: Working perfectly fine

File: server/plugin/api.go [hunk 3]
```diff
@@ -253,12 +253,22 @@ func (p *Plugin) createSubscription(w http.ResponseWriter, r *http.Request) {
...
 
-	if statusCode, err = client.CreateSubscription(subscription); err != nil {
+	resp, statusCode, err := client.CreateSubscription(subscription)
+	if err != nil {
 		_ = p.handleClientError(w, r, err, false, statusCode, "", "")
 		p.API.LogError("Error in creating subscription", "Error", err.Error())
 		return
 	}
 
+	if subscription != nil && subscription.SubscriptionNumber != nil {
```
[the remark this answers is not in the document]
Comment (Nityanand13) [reply server/plugin/api.go:263]: Not now... but for test cases written this.

File: server/plugin/command.go [hunk 4]
```diff
@@ -342,9 +343,13 @@ func (p *Plugin) handleDeleteSubscription(_ *plugin.Context, args *model.Command
 			return
 		}
 
-		if _, err = client.DeleteSubscription(subscriptionID); err != nil {
+		if statusCode, err := client.DeleteSubscription(subscriptionID); err != nil {
 			p.API.LogError("Unable to delete subscription", "Error", err.Error())
-			p.postCommandResponse(args, p.handleClientError(nil, nil, err, isSysAdmin, 0, args.UserId, ""))
+			if statusCode == http.StatusNotFound {
+				p.postCommandResponse(args, "Subscription is already deleted.")
```
[the remark this answers is not in the document]
Comment (Nityanand13) [reply server/plugin/command.go:349]: I don't think so

[the remark this answers is not in the document]
Comment (ayusht2810) [reply server/plugin/command.go:349]: why

File: server/serializer/subscription.go [hunk 5]
```diff
@@ -179,3 +182,84 @@ func SubscriptionFromJSON(data io.Reader) (*SubscriptionPayload, error) {
...
+func GetFormattedSubscriptionEvents(subscriptionEvents string) string {
+	formattedSubscriptionEvents := ""
+	events := strings.Split(subscriptionEvents, ",")
+	for index, event := range events {
+		formattedSubscriptionEvents += constants.FormattedEventNames[strings.TrimSpace(event)]
+		if index != len(events)-1 {
+			formattedSubscriptionEvents += ", "
+		}
+	}
+	return formattedSubscriptionEvents
```
Comment (ayusht2810) [inline server/serializer/subscription.go:195]:
```suggestion
	}
	
	return formattedSubscriptionEvents
```

File: server/constants/routes.go [hunk 6]
```diff
@@ -38,4 +38,6 @@ const (
 	PathSysUserGroup  = "/sys_user_group.do?sys_id=%s"
 	PathKnowledgeBase = PathServiceNowURL + "/kb_knowledge_base.do%%3Fsys_id=%s"
 	PathCategory      = PathServiceNowURL + "/kb_category.do%%3Fsys_id=%s"
+	PathRecordListURL = "%s/nav_to.do?uri=%s_list.do%%3Fsysparm_query=active=true"
+	PathRecordURL     = "%s/nav_to.do?uri=%s.do%%3Fsys_id=%s%%26sysparm_stack=%s_list.do%%3Fsysparm_query=active=true"
```
Comment (ayusht2810) [inline server/constants/routes.go:42]:
```suggestion
	PathRecordList = "%s/nav_to.do?uri=%s_list.do%%3Fsysparm_query=active=true"
	PathRecord     = "%s/nav_to.do?uri=%s.do%%3Fsys_id=%s%%26sysparm_stack=%s_list.do%%3Fsysparm_query=active=true"
```

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/client.go [hunk 7]
```diff
@@ -84,13 +84,13 @@ func (c *client) ActivateSubscriptions() (int, error) {
...
 
-func (c *client) CreateSubscription(subscription *serializer.SubscriptionPayload) (int, error) {
-	_, statusCode, err := c.CallJSON(http.MethodPost, constants.PathSubscriptionCRUD, subscription, nil, nil)
+func (c *client) CreateSubscription(subscription *serializer.SubscriptionPayload) (*serializer.SubscriptionResponse, int, error) {
+	subscriptionResult := &serializer.SubscriptionResult{}
+	_, statusCode, err := c.CallJSON(http.MethodPost, constants.PathSubscriptionCRUD, subscription, subscriptionResult, nil)
 	if err != nil {
-		return statusCode, errors.Wrap(err, "failed to create subscription in ServiceNow")
+		return nil, statusCode, errors.Wrap(err, "failed to create subscription in ServiceNow")
 	}
```
Comment (manojmalik20) [inline server/plugin/client.go:92]:
```suggestion
	}
	
```

File: server/plugin/client.go [hunk 8]
```diff
@@ -144,12 +144,13 @@ func (c *client) DeleteSubscription(subscriptionID string) (int, error) {
...
 
-func (c *client) EditSubscription(subscriptionID string, subscription *serializer.SubscriptionPayload) (int, error) {
-	_, statusCode, err := c.CallJSON(http.MethodPatch, fmt.Sprintf("%s/%s", constants.PathSubscriptionCRUD, subscriptionID), subscription, nil, nil)
+func (c *client) EditSubscription(subscriptionID string, subscription *serializer.SubscriptionPayload) (*serializer.SubscriptionResponse, int, error) {
+	subscriptionResult := &serializer.SubscriptionResult{}
+	_, statusCode, err := c.CallJSON(http.MethodPatch, fmt.Sprintf("%s/%s", constants.PathSubscriptionCRUD, subscriptionID), subscription, subscriptionResult, nil)
 	if err != nil {
-		return statusCode, errors.Wrap(err, "failed to update subscription from ServiceNow")
+		return nil, statusCode, errors.Wrap(err, "failed to update subscription from ServiceNow")
 	}
```
Comment (manojmalik20) [inline server/plugin/client.go:152]:
```suggestion
	}
	
```

File: server/serializer/subscription.go [hunk 9]
```diff
@@ -179,3 +182,85 @@ func SubscriptionFromJSON(data io.Reader) (*SubscriptionPayload, error) {
...
+func (s *SubscriptionResponse) CreateSubscriptionPost(botID, serviceNowURL string) *model.Post {
+	post := &model.Post{
+		ChannelId: s.ChannelID,
+		UserId:    botID,
+	}
+
+	subscriptionEvents := GetFormattedSubscriptionEvents(s.SubscriptionEvents)
+	titleLink := fmt.Sprintf(constants.PathRecordList, serviceNowURL, s.RecordType)
+	recordType := cases.Title(language.Und).String(s.RecordType)
+	postTitle := fmt.Sprintf("%s subscription created for [%s](%s)", constants.BulkSubscription, recordType, titleLink)
```
Comment (manojmalik20) [inline server/serializer/subscription.go:208]: This should be in the `else` of the below `if` block.

File: server/serializer/subscription.go [hunk 10]
```diff
@@ -179,3 +182,85 @@ func SubscriptionFromJSON(data io.Reader) (*SubscriptionPayload, error) {
...
+func (s *SubscriptionResponse) EditSubscriptionPost(botID, serviceNowURL string) *model.Post {
+	post := &model.Post{
+		ChannelId: s.ChannelID,
+		UserId:    botID,
+	}
+
+	subscriptionEvents := GetFormattedSubscriptionEvents(s.SubscriptionEvents)
+	recordType := cases.Title(language.Und).String(s.RecordType)
+	textLink := fmt.Sprintf(constants.PathRecordList, serviceNowURL, s.RecordType)
+	postText := fmt.Sprintf("%s subscription for [%s](%s)", constants.BulkSubscription, recordType, textLink)
```
Comment (manojmalik20) [inline server/serializer/subscription.go:241]: Same as above.

File: server/serializer/subscription.go [hunk 11]
```diff
@@ -179,3 +182,85 @@ func SubscriptionFromJSON(data io.Reader) (*SubscriptionPayload, error) {
...
+		formattedSubscriptionEvents += constants.FormattedEventNames[strings.TrimSpace(event)]
+		if index != len(events)-1 {
+			formattedSubscriptionEvents += ", "
+		}
+	}
+
+	return formattedSubscriptionEvents
+}
+
+func (s *SubscriptionResponse) CreateSubscriptionPost(botID, serviceNowURL string) *model.Post {
```
Comment (manojmalik20) [inline server/serializer/subscription.go:199]:
```suggestion
func (s *SubscriptionResponse) CreateSubscriptionCreatedPost(botID, serviceNowURL string) *model.Post {
```

File: server/serializer/subscription.go [hunk 12]
```diff
@@ -179,3 +182,85 @@ func SubscriptionFromJSON(data io.Reader) (*SubscriptionPayload, error) {
...
+				Value: subscriptionEvents,
+			},
+		},
+	}
+
+	model.ParseSlackAttachment(post, []*model.SlackAttachment{slackAttachment})
+	return post
+}
+
+func (s *SubscriptionResponse) EditSubscriptionPost(botID, serviceNowURL string) *model.Post {
```
Comment (manojmalik20) [inline server/serializer/subscription.go:232]:
```suggestion
func (s *SubscriptionResponse) CreateSubscriptionEditedPost(botID, serviceNowURL string) *model.Post {
```

File: server/serializer/subscription.go [hunk 13]
```diff
@@ -179,3 +182,85 @@ func SubscriptionFromJSON(data io.Reader) (*SubscriptionPayload, error) {
 
 	return sp, nil
 }
+
+func GetFormattedSubscriptionEvents(subscriptionEvents string) string {
```
Comment (manojmalik20) [inline server/serializer/subscription.go:186]: You can also use this function inside `GetFormattedSubscriptions` function. Also, take a reference from there for using `strings.Builder`.

File: server/serializer/servicenow_event.go [hunk 14]
```diff
@@ -79,15 +79,11 @@ func (se *ServiceNowEvent) CreateNotificationPost(botID, serviceNowURL, pluginUR
 		})
 	}
 
-	titleLink := fmt.Sprintf("%s/nav_to.do?uri=%s.do%%3Fsys_id=%s%%26sysparm_stack=%s_list.do%%3Fsysparm_query=active=true", serviceNowURL, se.RecordType, se.RecordID, se.RecordType)
+	titleLink := fmt.Sprintf(constants.PathRecord, serviceNowURL, se.RecordType, se.RecordID, se.RecordType)
 	slackAttachment := &model.SlackAttachment{
 		Title: fmt.Sprintf("[%s](%s): %s", se.Number, titleLink, se.ShortDescription),
+		Text:  fmt.Sprintf("**Event: %s**", constants.FormattedEventNames[se.EventOccurred]),
```
Comment (manojmalik20) [inline server/serializer/servicenow_event.go:85]: Please add screenshot for this in the PR description.

Comment (manojmalik20) [inline server/serializer/servicenow_event.go:85]: Please add screenshot for this in the PR description.

File: server/plugin/api_test.go [hunk 15]
```diff
@@ -1265,6 +1267,36 @@ func TestCreateSubscription(t *testing.T) {
 			ExpectedStatusCode:   http.StatusForbidden,
 			ExpectedErrorMessage: "create subscription error",
 		},
+		"failed to create the post": {
+			RequestBody: fmt.Sprintf(`{
+				"user_id": "%s",
+				"channel_id": "%s"
+			  	}`, testutils.GetID(), testutils.GetChannelID()),
```
Comment (manojmalik20) [inline server/plugin/api_test.go:1274]: You can create a testutil for this.

Comment (Nityanand13) [reply server/plugin/api_test.go:1274]: Can you please elaborate?

File: server/testutils/data.go [hunk 16]
```diff
@@ -12,6 +12,10 @@ import (
 	"github.com/mattermost/mattermost-plugin-servicenow/server/serializer"
 )
 
+func RequestBody() string {
```
Comment (manojmalik20) [inline server/testutils/data.go:15]:
```suggestion
func GetRequestBody() string {
```

**Review by manojmalik20 — APPROVED** [review]

**Review by manojmalik20 — APPROVED** [review]

**Comment by Nityanand13** [comment]
PR is made on the mattermost :[#169](https://github.com/mattermost/mattermost-plugin-servicenow/pull/169)

**Pull request closed without merging**